### ISSUE TITLE
MVP Agencies API -  Query and JSON changes

### ISSUE THREAD

**Issue by PhilNorman2** [midohiofoodbank/freshtrak-public#36]
After Initial Agencies API was completed @davidpickering , @karengreene, @nvengal and @PhilNorman2 discussed next steps for API.  They are described in this issue.

**Comment by PhilNorman2** [comment]
### Updated query.  Changes are in black 

David, Phil and Karen met and discussed the query and the following changes need to be made in the agency query:
(event_date_key >= submitted date thru API)
AND event_dates.published_date_key <= todays date

This makes the updated query updated as follows changes in bold

'event_date_key is the date that the event date occurs - this needs to be the date that is passed into the API
'published_date_key' is when it can be shown to public to limit events that have not been published
SELECT DISTINCT locations.* FROM locations
INNER JOIN foodbanks_mini ON foodbanks_mini.status_id = 1 AND foodbanks_mini.fb_id = locations.primary_fb_id
INNER JOIN foodbank_counties ON foodbank_counties.fb_id = foodbanks_mini.fb_id
INNER JOIN CNTY ON CNTY.FIPS = foodbank_counties.fips
INNER JOIN zip_codes ON zip_codes.priority = 1 AND zip_codes.fips = CNTY.FIPS
INNER JOIN events ON events.status_id = 1 AND events.status_publish_event = 1 AND events.loc_id = locations.loc_id INNER JOIN events events_locations_join ON events_locations_join.status_id = 1 AND events_locations_join.status_publish_event = 1 AND events_locations_join.loc_id = locations.loc_id
INNER JOIN event_dates ON event_dates.status_id = 1 AND event_dates.status_publish = 1 AND event_dates.event_id = events_locations_join.event_id AND (**event_date_key >= submitted date thru API**)
WHERE locations.status_id = 1 AND zip_codes.zip_code = '43030'
AND events.status_id = 1 AND events.status_publish_event = 1
AND events.status_publish_event_dates = 1
**AND event_dates.published_date_key <= todays date**

----
### **Update JSON will include address and distance for the events.** 

There is now an event_dates under events. Service is under events.
<pre>
{
&nbsp;&nbsp;"agencies":[
&nbsp;&nbsp;{
         "id":6,
          "address":"3960 BROOKHAM DR",
         "city":"GROVE CITY",
         "state":"OH",
         "zip":"43123",
         "phone":"614-317-9482",
         "name":"Mid-Ohio Foodbank - Kroger Community Pantry",
         "nickname":"MOF Kroger Pantry",
         "events":[
            {
               "name":"Mobile Pantry A",
               "address":"123 first address",
               "distance":"This is the distance to the 123 first address",
               "service":"Prepack Pantry",
               "event_dates":[
                  {
                     "id":1,
                     "start_time":"01:00 PM",
                     "end_time":"03:00 PM",
                     "date":"2020-04-06"
                  },
                  {
                     "id":2,
                     "start_time":"01:00 PM",
                     "end_time":"03:00 PM",
                     "date":"2020-04-08"
                  }
               ]
            },
            {
               "name":"Mobile Pantry B",
               "address":"456 different street",
               "distance":"This is the distance to the 123 first address",
               "service":"Prepack Pantry",
               "event_dates":[
                  {
                     "id":1,
                     "start_time":"01:00 PM",
                     "end_time":"03:00 PM",
                     "date":"2020-04-06"
                  },
                  {
                     "id":2,
                     "start_time":"01:00 PM",
                     "end_time":"03:00 PM",
                     "date":"2020-04-08"
                  }
               ]
            }
         ]
      }
   ]
}
</pre>

**Comment by PhilNorman2** [comment]
accidentally moved to done

**Comment by davidpickering** [comment]
Phil & Karen - Thank you both for a great conversation about this theme. At MOF we are very honored to have such a great group of people helping on this project. We couldn't do it without you all.

**Comment by davidpickering** [comment]
Okay I think I got my head around what we are trying to do and how it relates back to the data architecture. To start, at first glance the interpreted query coming out of Active Record seemed a little convoluted and redundant. However, I think I'm learning that Active Record is fundamentally different from just pure SQL and we would expect it to behave somewhat differently, but get us to the same result while giving us different tools and control.

For reference of people working on the project, here is the most straightforward way I would extract this data purely with SQL:

I would also like to outline a few concepts in the data structure that we all have continued to try and wrap our heads around:

**Logical Concept -** How the input of a zipcode relates to the agencies shown. 

**Issue**: Many zip codes cross a county boundary and it is more difficult to decide which agencies to show to a customer who inputs a zipcode. Example image shows Licking county, and all of its zip codes, note how many of them cross outside of the licking county boundary.

![image](https://user-images.githubusercontent.com/11682225/78624089-c8d2cc00-7856-11ea-89cb-2e56749faa0a.png)


User inputs a zipcode

![image](https://user-images.githubusercontent.com/11682225/78622757-fff3ae00-7853-11ea-87ef-92d0f9245a27.png)

Zipcode is looked  up in the zip_codes table. This table holds one entry per zipcode/county combination. If a single zip code crosses a county line, it will have two entries. If it is contained entirely within one county, it will only have one entry.

![image](https://user-images.githubusercontent.com/11682225/78622809-2580b780-7854-11ea-8287-5de96b09070a.png)


The fips codes from the resulting zipcode(s) are then joined against the zip_codes table again. We find all zipcodes that are a part of any of the returned counties (based on fips code)

![image](https://user-images.githubusercontent.com/11682225/78623030-b8215680-7854-11ea-8916-d73a6119f9e6.png)


Now look for any agencies (events.zip_code field) in any of those zip codes.
`
SELECT
	`events`.event_name,
	`events`.address1,
	`events`.address2,
	`events`.city,
	`events`.state,
	`events`.zip 
FROM
	`events`
	INNER JOIN zip_codes ON `events`.zip = zip_codes.zip_code
	INNER JOIN zip_codes AS zip_fips ON zip_codes.fips = zip_fips.fips 
WHERE
	zip_fips.zip_code = 43046
`

Those agencies are shown in the UI (by event_date, or in the "additional resources" section). Additional filtering applies based on target service geographies, but that will be covered in another issue.

![image](https://user-images.githubusercontent.com/11682225/78624230-249d5500-7857-11ea-889f-a4ac7439c9b0.png)

**Comment by karengreene** [comment]
Added city, state, zip to events structure in below comment. Missed adding it the first time. 
https://github.com/midohiofoodbank/freshtrak-public/issues/36#issuecomment-609998525

**Comment by karengreene** [comment]
Hi David,

Thanks for writing that up.  I executed your query from the comment 
https://github.com/midohiofoodbank/freshtrak-public/issues/36#issuecomment-610140143.  However, it does not bring back the same information as the below query even if I take out the filter for the event_date_key >= submitted date thru API) and  event_dates.published_date_key <= today's date.  The query you have returns 149 results. The below query only returns five results.

SELECT DISTINCT `locations`.* FROM `locations` 
INNER JOIN `foodbanks_mini` ON `foodbanks_mini`.`status_id` = 1 AND `foodbanks_mini`.`fb_id` = `locations`.`primary_fb_id` 
INNER JOIN `foodbank_counties` ON `foodbank_counties`.`fb_id` = `foodbanks_mini`.`fb_id` 
INNER JOIN `CNTY` ON `CNTY`.`FIPS` = `foodbank_counties`.`fips` 
INNER JOIN `zip_codes` ON `zip_codes`.`priority` = 1 AND `zip_codes`.`fips` = `CNTY`.`FIPS` 
INNER JOIN `events` ON `events`.`status_id` = 1 AND `events`.`status_publish_event` = 1 AND `events`.`loc_id` = `locations`.`loc_id` INNER JOIN `events` `events_locations_join` ON `events_locations_join`.`status_id` = 1 AND `events_locations_join`.`status_publish_event` = 1 AND `events_locations_join`.`loc_id` = `locations`.`loc_id` 
INNER JOIN `event_dates` ON `event_dates`.`status_id` = 1 AND `event_dates`.`status_publish` = 1 AND `event_dates`.`event_id` = `events_locations_join`.`event_id` 
WHERE `locations`.`status_id` = 1 AND `zip_codes`.`zip_code` = '43046' 
	AND `events`.`status_id` = 1 AND `events`.`status_publish_event` = 1 
    AND `events`.`status_publish_event_dates` = 1

**Comment by davidpickering** [comment]
> 
> 
> Hi David,
> 
> Thanks for writing that up. I executed your query from the comment
> [#36 (comment)](https://github.com/midohiofoodbank/freshtrak-public/issues/36#issuecomment-610140143). However, it does not bring back the same information as the below query even if I take out the filter for the event_date_key >= submitted date thru API) and event_dates.published_date_key <= today's date. The query you have returns 149 results. The below query only returns five results.
> 

@karengreene Yep, that is correct. The query I had in the above comment returns a different set of results than what is currently in production. I wanted to provide some more information about how to reference MOF's data sources to get the correct response. It seems like providing the conceptual pieces is more useful than providing just SQL since that has to be translated into Active Record objects.

In my above comment I was trying to outline the nuance between "find all agencies with x ZipCode" vs "Find all agencies in a county linked to any county that the input zip code relates to". I'm hoping we'll get the next iteration adapted to this logic.

**Comment by PhilNorman2** [comment]
Is this the problem we are trying to solve?

As a food seeker I want to see all events taking place in all of the zip
codes for the counties included in my zip code

**Comment by davidpickering** [comment]
@PhilNorman2 Yes! It sounds a little convoluted, but that is how we solve for borders and people who live at the edge of counties! 

More conceptual details in my above comment; https://github.com/midohiofoodbank/freshtrak-public/issues/36#issuecomment-610140143

**Comment by karengreene** [comment]
I created a card for this enhancement and put it in backlog.
